### DISCUSSION TITLE
How to use `NuxtLink` tag in post content from the server

### DISCUSSION THREAD

**Discussion by mirumirumi** [nuxt/nuxt#17059] [Questions]
In my blog site that built by Nuxt 3, I want to use `NuxtLink` tags in case of from my domain to my domain (equals internal links).

There is the content as a HTML in my DB, and this content inserted to my Nuxt component with `v-html` .

```html
      <article>
        <header>
          <h1>
            {{ post.title }}
          </h1>
        </header>
        <div id="content" v-html="post.body"></div>  // <- this
      </article>
```

However, Nuxt 3 doesn't parse `NuxtLink`, it becomes `<nuxtlink>`, a tag I don't understand, and the browser outputs it as is. In other words, nothing happens.

![image](https://user-images.githubusercontent.com/88871226/184617698-6c6213cc-d0e6-4782-8037-ace76e9ebaae.png)

I already understand that `v-html` does exec no Vue directive, so I also understand custom component such as `NuxtLink` is not working.

So, what I should do?

**Comment by amrnn90** [comment]
I am not sure if this is the best way, but try this:

Create a new component called `DynamicHtml.vue`:
```vue
<script>
import { h } from "vue";
import NuxtLink from "#app/components/nuxt-link"

export default {
  props: {
    html: {
      type: String,
    },
  },
  render() {
    return h({
      template: `<div>${this.html}</div>`,
      components: { NuxtLink }
    });
  },
};
</script>
```

Next, we need to have a Vue build that includes the template compiler (this will increase your app size!), there is an existing PR #4762 that will make this easier, but at this time you will need to install the following module:
https://github.com/huang-julien/nuxt-runtime-compiler

```bash
npm install -D nuxt-runtime-compiler
```

Edit `nuxt.config.ts`:
```ts
export default defineNuxtConfig({
  modules: ["nuxt-runtime-compiler"],
});
```

Finally you can use the component in your pages by passing the dynamic HTML to the `html` prop:

```vue
<template>
  <div>
    <DynamicHtml
      :html="`
        <h1>This is dynamic</h1>
        <NuxtLink href='/'>Go to Homepage</NuxtLink>
      `"
    />
  </div>
</template>

```

**Reply by mirumirumi** [reply]
Thanks for your reply.
I will report back on the results once I try it :)

**Comment by huang-julien** [comment]
Basically when you send any string into the v-html directive, vue does not parse your `<nuxtlink> ` as a component because the content is inserted as plain html. 
`NuxtLink` is a component.
Since you are retrieving data from your own DB, i believe you can make some changes to your db to adapt these data and create some components for them.

**Reply by mirumirumi** [reply]
Thanks for your reply.
I’m very sorry, but I don’t understand that you are talking about.

Since the content in the DB is already HTML, it would be difficult to parse this as `NuxtLink`.

You seem to be the one who suggested the workaround that @amrnn90 mentioned above, what are your own thoughts on it?

**Reply by huang-julien** [reply]
What i meant was that what you receive from your DB is considered as a string (like literally), not HTML. Vue will read your string then parse it to insert your string as HTML. However, since this is a v-HTML directive, your browser is not aware of Vue components (it might recognize web components tho, not sure about it). Using v-html is like using `yourNode.innerHTML = yourStringFromDB`. 
The workaround that @amrnn90 mentionned can be a solution, however, using the runtime compiler will increase you build size and might make you nuxt app a bit slower on cold starts (server-side) since the vue-server-render has to compile your string to render function then cache it. On client side, vue will also have to compile your string into render functions. So you have to consider that there's a lot of downside to using it.
Using the runtime-compiler is valid if you have a lot of components with dynamic template. If you have only this part of your app that use string to render its content (and some time), i would consider to adapt the DB because the runtime compiler is a bit overkill for this. If not, https://github.com/nuxt/framework/discussions/6630#discussioncomment-3399993 IS a solution.

**Reply by amrnn90** [reply]
Hello @mirumirumi ,
I agree with @huang-julien , using a runtime compiler for this case is overkill.
If you only need to handle links so that you can have the SPA experience for internal urls, then I suggest using normal anchor tags in your database html instead of `<NuxtLink>` and simply add an event listener to intercept when a user tries going through any of these links, you can try something like this:

```vue
<script setup>
function interceptLinks(e) {
  const linkEl = e.target.closest("a");
  if (!linkEl) return;

  const url = linkEl.getAttribute("href");
  if (!url.startsWith("/")) return;

  if (e.ctrlKey || e.shiftKey || e.altKey || e.metaKey) return;

  e.preventDefault();
  navigateTo(url);
}
</script>

<template>
  <div @click="interceptLinks" v-html="yourPostHtml" />
</template>
```

**Reply by mirumirumi** [reply]
@amrnn90 @huang-julien 
Thank you both kindly!
I was foolish for not realizing how simple it was to just handle the click event.
It worked without a problem.

**Reply by amrnn90** [reply]
You are welcome @mirumirumi
Please take another look at the code in my previous comment, I have added a conditional to skip the function if the user is holding a modifier key like `shift` or `ctrl` to allow them to open the link in another tab or window.
